### PR TITLE
Add support for different CPU time fields for SLURM

### LINKED ISSUE #160: Wrong SLURM field used for CpuDuration
From GGUS: [ICM](https://ggus.eu/index.php?mode=ticket_info&ticket_id=132922), [BelGrid-UCL](https://ggus.eu/index.php?mode=ticket_info&ticket_id=132925), and [UKI-LT2-QMUL](https://ggus.eu/index.php?mode=ticket_info&ticket_id=132929)

`slurm_acc.sh` currently uses `CPUTimeRAW` from `sacct`, which is elapsed time * number of CPUs. This results in Elapsed time being the same as the wall clock time.

What should be used is probably `TotalCPU`, which is the sum of the SystemCPU and UserCPU time used by the job. However, parsing this field is less trivial than `CPUTimeRAW` as it seems to use a format which is something like `dd-hh:mm:ss`.

### PR DESCRIPTION
**Pull request by tofu-rocketry**
Resolves #160.

This PR adds support for different time formats of CPU duration to be passed to the SLURM parser. This is to allow for the use of either `CPUTimeRAW` or `TotalCPU` from `sacct` which each have good and bad points depending on use case. I have decided not to enforce either format as the new field TotalCPU may not be so good for multicore jobs. **I've left it up to the user as to which they use.**

I've also added the different CPU time options to the `slurm_acc.sh` script, added unit tests, and tidied some whitespace.

### CODE REVIEW

File: apel/parsers/slurm.py [hunk 1]
```diff
@@ -63,6 +63,14 @@ def parse(self, line):
                               'NODE_FAIL', 'PREEMPTED', 'TIMEOUT'):
             return None
 
+        # Select CPU time parsting function based on field used.
```
Comment (kjopek) [inline apel/parsers/slurm.py]: Typo: parsing instead of parsting.

File: scripts/slurm_acc.sh [hunk 2]
```diff
@@ -1,8 +1,17 @@
...
 
+# This script can be run to extract usage accounting data from SLURM for later
+# parsing by an APEL client.
+
 sleep 2
 
 NOW=$(date +"%Y%m%d")
 FILE="/var/log/apel/slurm_acc.$NOW"
 
+# This is the original sacct call which uses CPUTimeRAW for CPU Duration.
```
Comment (gregcorbett) [inline scripts/slurm_acc.sh]: Is the intention here sites would use one `sacct` command or the other? If so, maybe a comment stating that would be helpful.

File: test/test_slurm.py [hunk 3]
```diff
@@ -45,6 +45,16 @@ def test_parse_line(self):
             ('297720.batch|batch|||2013-10-25T12:11:20|2013-10-25T12:11:36|00:00:16|16||1|1|wn37|3228K|23820K|COMPLETED'),
             ('321439.batch|batch|||2013-10-27T17:09:35|2013-10-28T04:47:20|11:37:45|41865||1|1|wn16|770728K|1.40G|COMPLETED'),
             ('320816.batch|batch|||2013-10-27T14:56:03|2013-10-28T05:03:50|14:07:47|50867||1|1|wn33|1325232K|2.22G|COMPLETED'),
+            # These lines use TotalCPU rather than CPUTimeRAW.
```
Comment (gregcorbett) [inline test/test_slurm.py]: So these new test cases make use of the `d-h:m:s` syntax as highlighted in bold?

> ('5940815|cream_877733810|user1|group1|2018-06-03T02:11:12|'
>  '2018-06-05T03:36:08|**2-01:24:56**|**8-15:40:36**|partition|8|1|'
>  'wn1|||COMPLETED'),

Comment (tofu-rocketry) [reply test/test_slurm.py]: Kinda. There should be a test for each new format: `d-h:m:s`, `h:m:s` and `m:s.s`

Comment (gregcorbett) [reply test/test_slurm.py]: ah yes, I found it quite hard to parse the test cases - maybe add a comment stating which new test case tests what format?

File: scripts/slurm_acc.sh (hunk 2 above)
Comment (gregcorbett) [reply scripts/slurm_acc.sh]: 👍

File: test/test_slurm.py (hunk 3 above)
Comment (tofu-rocketry) [reply test/test_slurm.py]: Done.

**Comment by tofu-rocketry** [comment]
I've rebased and squashed the extra commits into the earlier ones. Could I get a review from someone?

File: apel/common/datetime_utils.py [hunk 4]
```diff
@@ -58,14 +58,19 @@ def parse_timestamp(datetime_string):
...
-    Return seconds from times of the form d-h:m:s or h:m:s.
-    '''
+    """Return integer seconds from times of form d-h:m:s, h:m:s or m:s.s."""
     if '-' in timestring:
         days, sub_days = timestring.split('-')
     else:
         days, sub_days = 0, timestring
-    hours, minutes, seconds = sub_days.split(':')
+
+    if '.' in sub_days:
```
Comment (jrha) [inline apel/common/datetime_utils.py]: Maybe I just like regexps too much, but I'd probably try to parse the times with something like:
`(?:(?P<days>\d)-)?(?:(?P<hours>\d{2}):)?(?P<minutes>\d{2}):(?P<seconds>\d{2}(?:\.\d+)?)`

Although... maybe you could loop over a series of `strptime()` formats, but there would be problems with day numbering and fractional seconds... nope, regexps FTW.

Comment (tofu-rocketry) [reply apel/common/datetime_utils.py]: Feel free to submit a PR! :wink: 

It's a fair point, but I wanted to fix the current problem rather than introduce a new factor in.
